lib/rpmsg/rpmsg_virtio.c

```diff
@@ -432,6 +432,10 @@ static void rpmsg_virtio_rx_callback(struct virtqueue *vq)
 
 		rp_hdr = (struct rpmsg_hdr *)
 			 rpmsg_virtio_get_rx_buffer(rvdev, &len, &idx);
+		if (rp_hdr == NULL) {
+			/* tell peer we return some rx buffer */
+			virtqueue_kick(rvdev->rvq);
```

Comment:
Two case here:
1.Linux implementation enable rx callback only when it can't get the buffer from the virtio queue and will disable rx callback in all other condition.
Please reference rpmsg_upref_sleepers/rpmsg_downref_sleepers here:
https://github.com/torvalds/linux/blob/master/drivers/rpmsg/virtio_rpmsg_bus.c
2.OpenAMP use the busy loop to wait the buffer available, so we should always disable rx callback uncoditionally, please see this pull request:
https://github.com/OpenAMP/open-amp/pull/147/commits/370b5692b45607fb2826f4895243f25701528d5b
So the kick only happen when Linux is waiting for the buffer.

Reply:
@arnop2 this patch isn't optimizition, it's a bug fix to avoid kernel side stuck to wait for rx callback.

Reply:
@xiaoxiang781216 You right, this fix is needed, to avoid that Linux is stuck, waiting free buffers.
so i'm ok with the fix, even if it generates extra interrupt in Linux when there is no more message to treat. So potentially for each message sent by Linux. 
I need to go deep in vring flag management (not fluent with this part...) to review 370b569. Anyway is it a way to not introduce the master/slave role notion in it? i will expect more a sender/receiver view for the vrings.

Reply:
If some free buffer is available in queue, kernel(rpmsg_downref_sleepers) will disable the callback immediately by set VRING_USED_F_NO_NOTIFY flag, so no extra interrupt happen here. If we trigger the interrupt because VRING_USED_F_NO_NOTIFY isn't set, which mean kernel really need this interrupt to wake up.

Reply:
To work correctly, vring need to know virtio device contain it is slave or master, since vring need check/modify the different field for interrupt management. The detailed difference could reference this pull request:
https://github.com/OpenAMP/open-amp/pull/147